### PR TITLE
feat(stepfunctions-tasks): support for the Step Functions optimized integration for Bedrock InvokeModel API

### PR DESCRIPTION
**Pull request by clareliguori**
Step Functions recently released an optimized integration for Bedrock InvokeModel API, and these changes add support for adding the Bedrock InvokeModel task to Step Functions state machines.

Closes https://github.com/aws/aws-cdk/issues/28268.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by scanlonp** [comment]
Hey @clareliguori, this looks terrific! Before I dive into a full review, I had a couple of higher level questions. I see you created a `aws-cdk-lib/aws-bedrock` folder, but that it does not correspond to any cloudformation resources. You explain that in the readme, but I am concerned this could be confusing for users when they see an import from a bedrock module. Are there use cases for the bedrock models outside of stepfunctions? Or is there a reason they should not be contained in the stepfunctions module? Are there other cases in stepfunctions that have had a similar pattern?

I will also ask the team about if we have dealt with patterns similar to this, and how we approach them.

Also a quick note about the bedrock output. Did you notice Noah and Oliver are out of alphabetical order! That is pretty funny 🙃

**Comment by clareliguori** [comment]
> Hey @clareliguori, this looks terrific! Before I dive into a full review, I had a couple of higher level questions. I see you created a `aws-cdk-lib/aws-bedrock` folder, but that it does not correspond to any cloudformation resources. You explain that in the readme, but I am concerned this could be confusing for users when they see an import from a bedrock module. Are there use cases for the bedrock models outside of stepfunctions? Or is there a reason they should not be contained in the stepfunctions module? Are there other cases in stepfunctions that have had a similar pattern?
> 
> I will also ask the team about if we have dealt with patterns similar to this, and how we approach them.

Yeah I went back and forth on that, and I'm happy to change it

Pros of being in the aws-bedrock module:
* These import-only constructs can be expanded to add IAM permission-type methods, like grantInvokeModel, in order to grant permissions to invoke the models to other constructs like ECS services, Lambda functions, etc.
* If CloudFormation does eventually support a resource like Bedrock ProvisionedThroughput (I don't have any insight on if they will or not), then it would be easier to expand on the existing class for the L2 construct vs having some duplicate and/or deprecated logic between the aws-stepfunctions-tasks and aws-bedrock modules.

Cons of being in the aws-bedrock module:
* Since we don't know what any potential future Bedrock resource type would look like, I'm kind of guessing at the right shape/naming of the L2 construct with these initial classes. We could end up deprecating one or more of the classes anyway if I got it wrong.

Note that I don't expect a base foundational model would ever be a CFN resource (since they don't belong to an account), so that would likely always be an import-only construct as it is in this PR.

> Also a quick note about the bedrock output. Did you notice Noah and Oliver are out of alphabetical order! That is pretty funny 🙃

LOL I did not notice that

**Review by kaizencc — DISMISSED** [review]
Hi @clareliguori, thanks for the PR! This is an interesting one, and my feelings are that, while unusual, your `aws-bedrock` pre-CFN implementation provides some benefit in creating the model arn, and is relatively low-risk in predicting what a bedrock L2 would look like in the future.

Know that our normal process for creating new L2s involves an RFC, experimental module first, and much more bureaucracy. But I think common sense should prevail here as this is simple enough.

Sending this off to @rix0rrr to glance at just to make sure there's not something obvious i'm missing about creating the `aws-bedrock` submodule without cfn. I doubt that will cause any issues, although when bedrock gets CFN resources we'll have to rely on our L1 generation to be resilient enough to play nice with the existing folder.

File: packages/aws-cdk-lib/aws-bedrock/lib/foundation-model.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import { Construct } from 'constructs';
+import { IModel } from './model-base';
+import { ArnFormat, Stack } from '../../core';
+
+/**
+ * The model identifiers for the Bedrock base foundation models.
+ *
+ * @see https://docs.aws.amazon.com/bedrock/latest/userguide/model-ids-arns.html
+ */
+export enum FoundationModelIdentifier {
```
[the remark this answers is not in the document]
Comment (clareliguori) [reply packages/aws-cdk-lib/aws-bedrock/lib/foundation-model.ts:10]: Or alternatively, make these all `public static readonly` strings within the FoundationModel class, and change the `fromFoundationModelId` method to just take a string?
```
public static fromFoundationModelId(scope: Construct, _id: string, foundationModelId: string): FoundationModel
```

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/bedrock/invoke-model.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+ * Properties for invoking a Bedrock Model
+ */
+export interface BedrockInvokeModelProps extends sfn.TaskStateBaseProps {
+
+  /**
+   * The Bedrock model that the task will invoke.
+   *
+   * @see https://docs.aws.amazon.com/bedrock/latest/userguide/api-methods-run.html
+   */
+  readonly model: bedrock.IModel;
```
[the remark this answers is not in the document]
Comment (clareliguori) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/bedrock/invoke-model.ts:54]: I think the most common case will be using foundation models, so having that ARN constructed for you is pretty useful

File: packages/aws-cdk-lib/aws-bedrock/lib/foundation-model.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaizencc) [reply packages/aws-cdk-lib/aws-bedrock/lib/foundation-model.ts:10]: I fear that may get unruly as we continue to add more `static readonly` strings to `FoundationModel`. In practice, our users tend to like the provided methods over inputing the string themselves even though they're the exact same thing.

The enum-like class pattern is typically what we do for scenarios like this (i.e. lambda runtimes).

**Review by kaizencc — DISMISSED** [review]
Looks good! From my understanding of bedrock, it looks like at least the foundation model will never be a CFN resource since we are referencing something that exists rather than creating something new. So the bedrock pseudo-construct is actually the best we can do and will be useful if there are other places we want to send in a bedrock arn. Thanks @clareliguori

**Comment by clareliguori** [comment]
@kaizencc I need your approval again - I ran into this issue with Mergify:
https://github.com/Mergifyio/mergify/issues/5055

**Review by kaizencc — APPROVED** [review]

**Comment by linkcd** [comment]
Thanks for this PR! Sorry for a basic question, but how do i check which CDK version will contain this RP?

**Comment by clareliguori** [comment]
@linkcd  This is now released in 2.115.0
https://github.com/aws/aws-cdk/releases/tag/v2.115.0

**Pull request merged**